homeassistant/components/roon/server.py

```diff
@@ -0,0 +1,156 @@
+"""Code to handle the api connection to a Roon server."""
+from asyncio import ensure_future, run_coroutine_threadsafe, sleep
+import logging
+
+from roon import RoonApi
+
+from homeassistant.const import CONF_API_KEY, CONF_HOST
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.util.dt import utcnow
+
+from .const import ROON_APPINFO
+
+_LOGGER = logging.getLogger(__name__)
+FULL_SYNC_INTERVAL = 30
+
+
+class RoonServer:
+    """Manages a single Roon Server."""
+
+    def __init__(self, hass, config_entry):
+        """Initialize the system."""
+        self.config_entry = config_entry
+        self.hass = hass
+        self.roonapi = None
+        self.all_player_ids = []
+        self.all_playlists = []
+        self.offline_devices = []
+        self._exit = False
+
+    @property
+    def host(self):
+        """Return the host of this server."""
+        return self.config_entry.data[CONF_HOST]
+
+    async def async_setup(self, tries=0):
+        """Set up a roon server based on host parameter."""
+        host = self.host
+        hass = self.hass
+        token = self.config_entry.data[CONF_API_KEY]
+        _LOGGER.debug("async_setup: %s %s", token, host)
+        self.roonapi = RoonApi(ROON_APPINFO, token, host, blocking_init=False)
+        self.roonapi.register_state_callback(
+            self.roonapi_state_callback, event_filter=["zones_changed"]
+        )
+
+        # initialize media_player platform
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(
+                self.config_entry, "media_player"
+            )
+        )
+
+        # Initialize Roon background polling
+        ensure_future(self.async_do_loop())
+
+        return True
+
+    async def async_reset(self):
+        """Reset this connection to default state.
+
+        Will cancel any scheduled setup retry and will unload
+        the config entry.
+        """
+        self.stop_roon()
+        return True
+
+    @property
+    def zones(self):
+        """Return list of zones."""
+        return self.roonapi.zones
+
+    def stop_roon(self):
+        """Stop background worker."""
+        self.roonapi.stop()
+        self._exit = True
+
+    def roonapi_state_callback(self, event, changed_zones):
+        """Callbacks from the roon api websockets."""
+        run_coroutine_threadsafe(
```

Comment:
If we don't need to wait for the result of the task we can use `hass.add_job` instead.